# Hyperparameter Tuning using HyperDrive

TODO: Import Dependencies. In the cell below, import all the dependencies that you will need to complete the project.

In [1]:
import logging
import os
import csv

from matplotlib import pyplot as plt
import numpy as np
import pandas as pd
import pkg_resources

import azureml.core
from azureml.core import Environment, ScriptRunConfig
from azureml.core.experiment import Experiment
from azureml.core.workspace import Workspace
from azureml.core.dataset import Dataset


from azureml.widgets import RunDetails
from azureml.train.hyperdrive.run import PrimaryMetricGoal
from azureml.train.hyperdrive.policy import BanditPolicy
from azureml.train.hyperdrive.sampling import RandomParameterSampling
from azureml.train.hyperdrive.runconfig import HyperDriveConfig
from azureml.train.hyperdrive.parameter_expressions import uniform, choice
import shutil

# Check core SDK version number
print("SDK version:", azureml.core.VERSION)

SDK version: 1.27.0


In [2]:
ws = Workspace.from_config()

experiment_name = 'hyperdrive-run'
project_folder = './pipeline-project'

experiment=Experiment(ws, experiment_name)
run = experiment.start_logging()

In [3]:
from azureml.core.compute import AmlCompute
from azureml.core.compute import ComputeTarget
from azureml.core.compute_target import ComputeTargetException

# NOTE: update the cluster name to match the existing cluster
# Choose a name for your CPU cluster
amlcompute_cluster_name = "capstone-cluster"

# Verify that cluster does not exist already
try:
    compute_target = ComputeTarget(workspace=ws, name=amlcompute_cluster_name)
    print('Found existing cluster, use it.')
except ComputeTargetException:
    compute_config = AmlCompute.provisioning_configuration(vm_size='STANDARD_D2_V2',# for GPU, use "STANDARD_NC6"
                                                           #vm_priority = 'lowpriority', # optional
                                                           max_nodes=4)
    compute_target = ComputeTarget.create(ws, amlcompute_cluster_name, compute_config)

compute_target.wait_for_completion(show_output=True, min_node_count = 1, timeout_in_minutes = 10)
# For a more detailed view of current AmlCompute status, use get_status().

Found existing cluster, use it.
Succeeded
AmlCompute wait for completion finished

Minimum number of nodes requested have been provisioned


In [4]:
from azureml.core import Environment

ml_env = Environment.from_conda_specification(name = 'ml-env', file_path = './dependencies.yml')

## Hyperdrive Configuration

TODO: Explain the model you are using and the reason for chosing the different hyperparameters, termination policy and config settings.

In [5]:
# Specify parameter sampler
ps = RandomParameterSampling( {  
    "C": choice(0.001, 0.01, 0.1, 1, 10, 20, 50, 100),
    "max_iter": choice(50, 75, 100,), 
    "solver" : choice('liblinear', 'sag', 'lbfgs', 'saga')
    }
)

# Specify a Policy: check the job every 2 iterations and if the primary metric falls outside the top 50% range, 
# Azure ML should terminate the job
policy = BanditPolicy(evaluation_interval=2, slack_factor=0.2)


est = ScriptRunConfig(
    source_directory='.', 
    compute_target=compute_target, 
    script='train.py',
    environment=ml_env)

# Create a HyperDriveConfig using the estimator, hyperparameter sampler, and policy.
hyperdrive_config = HyperDriveConfig(
    run_config=est, 
    hyperparameter_sampling=ps,
    primary_metric_name="AUC_weighted", 
    primary_metric_goal=PrimaryMetricGoal.MAXIMIZE,
    policy=policy,
    max_total_runs=10, 
    max_concurrent_runs=5)

In [8]:
hypdrive_run = experiment.submit(hyperdrive_config)

## Run Details

OPTIONAL: Write about the different models trained and their performance. Why do you think some models did better than others?

TODO: In the cell below, use the `RunDetails` widget to show the different experiments.

In [9]:
RunDetails(hypdrive_run).show()
hypdrive_run.wait_for_completion(show_output=True)

_HyperDriveWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO'…

RunId: HD_359746e7-73a3-46ef-90ea-02fc460a9dd7
Web View: https://ml.azure.com/runs/HD_359746e7-73a3-46ef-90ea-02fc460a9dd7?wsid=/subscriptions/9e65f93e-bdd8-437b-b1e8-0647cd6098f7/resourcegroups/aml-quickstarts-144967/workspaces/quick-starts-ws-144967&tid=660b3398-b80e-49d2-bc5b-ac1dc93b5254

Streaming azureml-logs/hyperdrive.txt

"<START>[2021-05-16T14:35:31.479566][API][INFO]Experiment created<END>\n""<START>[2021-05-16T14:35:33.059421][GENERATOR][INFO]Trying to sample '5' jobs from the hyperparameter space<END>\n""<START>[2021-05-16T14:35:33.244279][GENERATOR][INFO]Successfully sampled '5' jobs, they will soon be submitted to the execution target.<END>\n"

Execution Summary
RunId: HD_359746e7-73a3-46ef-90ea-02fc460a9dd7
Web View: https://ml.azure.com/runs/HD_359746e7-73a3-46ef-90ea-02fc460a9dd7?wsid=/subscriptions/9e65f93e-bdd8-437b-b1e8-0647cd6098f7/resourcegroups/aml-quickstarts-144967/workspaces/quick-starts-ws-144967&tid=660b3398-b80e-49d2-bc5b-ac1dc93b5254



{'runId': 'HD_359746e7-73a3-46ef-90ea-02fc460a9dd7',
 'target': 'capstone-cluster',
 'status': 'Completed',
 'startTimeUtc': '2021-05-16T14:35:31.25048Z',
 'endTimeUtc': '2021-05-16T14:41:04.638324Z',
 'properties': {'primary_metric_config': '{"name": "AUC_weighted", "goal": "maximize"}',
  'resume_from': 'null',
  'runTemplate': 'HyperDrive',
  'azureml.runsource': 'hyperdrive',
  'platform': 'AML',
  'ContentSnapshotId': '6a2e26b8-b9f2-47ed-b669-c2b00e50dc16',
  'score': '0.8245761162898213',
  'best_child_run_id': 'HD_359746e7-73a3-46ef-90ea-02fc460a9dd7_4',
  'best_metric_status': 'Succeeded'},
 'inputDatasets': [],
 'outputDatasets': [],
 'logFiles': {'azureml-logs/hyperdrive.txt': 'https://mlstrg144967.blob.core.windows.net/azureml/ExperimentRun/dcid.HD_359746e7-73a3-46ef-90ea-02fc460a9dd7/azureml-logs/hyperdrive.txt?sv=2019-02-02&sr=b&sig=pljnC0j3O9fkZhkTpMRsk7f%2FzOHB1R4H%2FwQdgo7Oasg%3D&st=2021-05-16T14%3A31%3A10Z&se=2021-05-16T22%3A41%3A10Z&sp=r'},
 'submittedBy': 'ODL_User 1

## Best Model

TODO: In the cell below, get the best model from the hyperdrive experiments and display all the properties of the model.

In [12]:
import joblib
# Get your best run and save the model from that run.

hd_best_run = hypdrive_run.get_best_run_by_primary_metric()
hd_best_model = hd_best_run.register_model(model_name="hd_best_model" , model_path="./")
parameter_values = hd_best_run.get_details()

print("Best Run Id: ", hd_best_run.id)


best_run_metrics = hd_best_run.get_metrics() 

print("Best run metrics: ", best_run_metrics)
print("AUC_weighted: ", best_run_metrics["AUC_weighted"])


Best Run Id:  HD_359746e7-73a3-46ef-90ea-02fc460a9dd7_4
Best run metrics:  {'Regularization Strength:': 1.0, 'Max iterations:': 75, 'Algorithm:': 'liblinear', 'AUC_weighted': 0.8245761162898213}
AUC_weighted:  0.8245761162898213


## Model Deployment

Since the hyperparameter optimization did not result in a better, I decided to deploy the AutoML model as a web service. 
The best AUC_weighted metric for a HD model was: 0.8245761162898213

In [13]:
print("Best run details ", parameter_values)
print("Parameters: ", parameter_values["runDefinition"]["arguments"])

Best run details  {'runId': 'HD_359746e7-73a3-46ef-90ea-02fc460a9dd7_4', 'target': 'capstone-cluster', 'status': 'Completed', 'startTimeUtc': '2021-05-16T14:36:42.392088Z', 'endTimeUtc': '2021-05-16T14:37:30.710408Z', 'properties': {'_azureml.ComputeTargetType': 'amlcompute', 'ContentSnapshotId': '6a2e26b8-b9f2-47ed-b669-c2b00e50dc16', 'ProcessInfoFile': 'azureml-logs/process_info.json', 'ProcessStatusFile': 'azureml-logs/process_status.json'}, 'inputDatasets': [], 'outputDatasets': [], 'runDefinition': {'script': 'train.py', 'command': '', 'useAbsolutePath': False, 'arguments': ['--C', '1', '--max_iter', '75', '--solver', 'liblinear'], 'sourceDirectoryDataStore': None, 'framework': 'Python', 'communicator': 'None', 'target': 'capstone-cluster', 'dataReferences': {}, 'data': {}, 'outputData': {}, 'jobName': None, 'maxRunDurationSeconds': 2592000, 'nodeCount': 1, 'priority': None, 'credentialPassthrough': False, 'identity': None, 'environment': {'name': 'ml-env', 'version': 'Autosave_20